classes/class2/README.md

```diff
@@ -1 +1,303 @@
+
+## Functions
+
+Functions make complexity manageable! JavaScript is all about functions. Let's make some.
+
+At this point we should start putting our code in files. If you have a JavaScript file:
+
+###### functions.js
+```node
+function func1() {
+	return 1
+}
+```
+
+You can run it like this:
+
+```bash
+❯❯❯ node functions.js
+```
+
+In `functions.js`, we declare a function named `func1` that takes no arguments and returns the number `1`.
+
+We can add to `functions.js` to learn more about `func1`.
+
+###### functions.js
+```node
+...
+console.log( func1 )
+console.log( func1() )
+```
+
+The function `console.log` does what you might expect -- it logs arguments to the console.
+
+``` bash
+❯❯❯ node functions.js
+[Function: func1]
+1
+```
+
+The two print statements we added to `functions.js` look almost the same, but the console output shows that they're very different. The first statement prints the function object `func1` -- that is, the actual *entity* that represents the function itself (the code, the arguments, the scope in which it was declared -- more on scope later). Node says that `func1` is a `Function` named `func1` -- nothing unexpected here.
+
+Putting `()` next to the name of a function calls the function, so when we print `func1()` we get the return value of `func1`, which is `1`.
+
+`()` will actually always attempt to "call" the expression to its left, treating it like a function. So you can get errors like this:
+
+```node
+> 5()
+TypeError: number is not a function
+```
+
+A variable can also store a function declaration, so we could have done this instead:
+
+```node
+var func1 = function() {
+	return 1
+}
+```
+
+The above example wouldn't have acted *quite* the same, though, which brings us to what makes JavaScript functions so special.
+
+### Functions are Objects
+
+The print statements above hinted at this: in JavaScript, functions are objects. So something like this is possible:
+
+###### functions.js
+```node
+function func1() {
+	var func2 = function() {
+		return 2
+	}
+	return func2
+}
+```
+
+What will happen when we call `func1`?
+
+###### functions.js
+```node
+...
+console.log( func1() )
+```
+
+```bash
+❯❯❯ node functions.js
+[Function]
+```
+
+The function `func1` is a named function that returns the variable `func2`. We set `func2` equal to a function definition without a name -- the function on the right hand side of this assignment is known as an **anonymous function**.
+
+So, `func1` returns an anonymous function which returns `2`. How do we get at the return value of the returned function?
+
+###### functions.js
+```node
+...
+console.log( func1()() )
+```
+
+```bash
+❯❯❯ node functions.js
+2
+```
+
+This is totally legal. It might help to break down the expression we just printed:
+
+```node
+func1()() // let's add parentheses to show the order of evaluation
+( func1() )() // first, func1 is called
```

Comment:
I guess this could be a confusing example — do you have any ideas as to how to make it clearer? I was thinking maybe omitting the second line and changing the explanation.